docs/Fleet management Design.md

```diff
@@ -0,0 +1,163 @@
+is in the push model.
+
+However, in order to differentiate from the existing pull based model, we will implement a new controller to watch the `ManagedCluster` custom resources.
+In this way, we can start our project with a clean sheet but still keep compatible with most the existing OCM solutions. 
+
+### API Design
+Just to make this document more self-contained. Here I paste the golang definition of the `ManagedCluster` CRD spec.
+
+```golang
+
+type ManagedClusterSpec struct {
+	// +optional
+	ManagedClusterClientConfigs []ClientConfig `json:"managedClusterClientConfigs,omitempty"`
+
+	// +required
+	HubAcceptsClient bool `json:"hubAcceptsClient"`
+
+	// +optional
+	LeaseDurationSeconds int32 `json:"leaseDurationSeconds,omitempty"`
+}
+
+// TODO: we can add the cert information here so we don't need a kubeconfig
+type ClientConfig struct {
+    // URL is the URL of apiserver endpoint of the managed cluster.
+    // +required
+    URL string `json:"url"`
+
+    // CABundle is the ca bundle to connect to apiserver of the managed cluster.
+    // System certs are used if it is not set.
+    // +optional
+    CABundle []byte `json:"caBundle,omitempty"`
+}
+
+// ManagedClusterStatus represents the current status of joined managed cluster.
+type ManagedClusterStatus struct {
+    // Conditions contains the different condition statuses for this managed cluster.
+    Conditions []metav1.Condition `json:"conditions"`
+    
+    // Capacity represents the total resource capacity from all nodeStatuses
+    // on the managed cluster.
+    Capacity ResourceList `json:"capacity,omitempty"`
+    
+    // Allocatable represents the total allocatable resources on the managed cluster.
+    Allocatable ResourceList `json:"allocatable,omitempty"`
+    
+    // Version represents the kubernetes version of the managed cluster.
+    Version ManagedClusterVersion `json:"version,omitempty"`
+    
+    // +optional
+    ClusterClaims []ManagedClusterClaim `json:"clusterClaims,omitempty"`
+}
+
+type ManagedClusterClaim struct {
+    // Name is the name of a ClusterClaim resource on managed cluster. It's a well known
+    // or customized name to identify the claim.
+    Name string `json:"name,omitempty"`
+    
+    // Value is a claim-dependent string
+    Value string `json:"value,omitempty"`
+}
+
+// ResourceList defines a map for the quantity of different resources, the definition
+// matches the ResourceList defined in k8s.io/api/core/v1.
+type ResourceList map[ResourceName]resource.Quantity
+
+```
+
+### Design details
+Here is how we handle the following functionalities in the OCM in our new controller. The overall idea is that 
+a `managedCluster` CR represents a member cluster. A cluster joins the fleet when a corresponding `managedCluster` CR is
+successfully applied to the hub cluster. Similarly, a member cluster leaves the hub cluster when its corresponding 
+`managedCluster` CR is removed from the cluster. 
+
+#### Join a member cluster to the fleet
```

Comment:
an alternative approach is this controller on hub use the kubeconfig of spoke to deploy the registration/work agent on the spoke with the bootstrap secret. Would you compare the two approach?

Reply:
Do we still need the registration controller after this? Does the work agent still need to pull from the hub cluster in this case? I am thinking of making this a pure push model.

Reply:
it might not need registration process for cluster, but i think you probably still need registration process for addon. For example, an addon that needs to run on the spoke but needs to talk to hub apiserver.